File: 30-Day LeetCoding Challenge/Week 3 April 15th–April 21st /Number of Islands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        
        if len(grid) < 1:
            return 0
        
        def dfs (r,c):
            if r<0 or c<0 or r>=rows or c>=col or grid[r][c] == '0':
                return;
            grid[r][c] = '0'

            dfs(r-1,c)
            dfs(r+1,c)
            dfs(r,c+1)
            dfs(r,c-1)

        rows = len(grid)
        col = len(grid[0])

        island = 0

        for i in range(rows):
            for j in range(col):

                if grid[i][j] == '1':
                    island += 1
                    dfs(i,j)
        return island
           
            
        

File: 30-Day LeetCoding Challenge/Week 3 April 15th–April 21st /Number of Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        
        int x = grid.size();
        int y = 0;
        if(x>0)
            y = grid[0].size();
        int ans = 0;
        for(int i=0;i<x;i++){
            for(int j=0;j<y;j++){
                if(grid[i][j] == '1')
                {
                    ans += 1;
                    findisland(i,j,grid);
                }
            }
        }
        return ans;
    }
private:
    void findisland(int a, int b, vector<vector<char>>& grid) {
        
        int x = grid.size();
        int y = grid[0].size();
        
        if(a<0 or b<0 or a>=x or b>=y or grid[a][b] != '1')
            return;
        
        int xaxis[] = {-1, 0, 0, 1};
        int yaxis[] = {0, 1, -1, 0};
        
        grid[a][b] = '0';
        
        for(int i=0; i< 4; i++){
            findisland(a+xaxis[i], b+yaxis[i], grid);
        }
        
    }
    
};